bootstrap/kubeadm/internal/cloudinit/node_test.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloudinit
+
+import (
+	"fmt"
+	"testing"
+
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha4"
+	"sigs.k8s.io/yaml"
+)
+
+func TestNewNode(t *testing.T) {
+	tests := []struct {
+		name    string
+		input   *NodeInput
+		check   func([]byte) error
+		wantErr bool
+	}{
+		{
+			"check for duplicated WriteFiles",
+			&NodeInput{
+				BaseUserData: BaseUserData{
+					AdditionalFiles: []bootstrapv1.File{
+						{
+							Path:        "/etc/foo.conf",
+							Content:     "bar",
+							Owner:       "root",
+							Permissions: "0644",
+						},
+					},
+				},
+			},
+			checkWriteFiles("/etc/foo.conf", "/run/kubeadm/kubeadm-join-config.yaml", "/run/cluster-api/placeholder"),
+			false,
+		},
+		{
+			"check for existens of /run/kubeadm/kubeadm-join-config.yaml and /run/cluster-api/placeholder",
+			&NodeInput{},
+			checkWriteFiles("/run/kubeadm/kubeadm-join-config.yaml", "/run/cluster-api/placeholder"),
```

Comment:
I would prefer adding the files here directly and inlining the check code in l.65+.

It's imho far more common and easier to understand. If I got it correctly they are something like `wantFiles` or `expectedFiles`

Reply:
+ I think it easier to read if the fields are explicitly added in the struct literal

Reply:
The idea behind the test func for me was:
Checking for duplicate Files is a single case and the NewNode function may have others aspects to Test. These other aspects could geht implemented by other functions. Also Not inlining it makes it also usable for the join controlplane equivalent func.

But I'm also OK with inlining it.

Reply:
Ah got it. Now it makes sense to me. Fine to keep as is for me.